design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,219 @@
+For example, let’s consider this VM spec:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+kubelet will emit following event:
+```
+SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
+```
+The following picture helps show how this is a problem
+<img src="smt-slignment-error-diagram.png">
+
+## Goals
+Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.
+
+## Non Goals
+Set the virt-launcher Pod scheduling to match a node with SMT aligned to its request.
+
+## Definition of Users
+This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.
+
+## User Stories
+* As a cluster admin, I want to enable VMs with an even number of CPUs and `isolatedEmulatorThread` enabled to run on my cluster by setting a cluster-wide knob on HCO.
+* As a Kubevirt VM owner, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)
+
+# Bug fix design proposal
+The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadCompleteToEvenParity`:
+```
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  annotations:
+    kubevirt.io/EmulatorThreadCompleteToEvenParity:
+spec:
+  configuration:
+    developerConfiguration:
+      featureGates:
+      - EmulatorThreadCompleteToEvenParity
+```
+This annotation will be respected only when the `EmulatorThreadCompleteToEvenParity` featureGate is enabled.
+
+## Virt-API:
+
+### VMI mutator webhook
+When the following conditions are met:
+1. The feature gate is enabled on the Kubevirt CR
+2. Kubevirt CR has the annotation.
+3. VMI is configured with `isolateEmulatorThread: true`
+
+then the VMI mutator webhook will copy the annotation from the kubevirt CR to the VMI annotations. This is done in order to pass the information between the virt-controller, virt-handler and virt-launcher, without changing the API between the virt-handler and the virt-launcher. Doing so avoids adding API on the virt-handler to virt-launcher grpc protocol.
+
+### VMI create-admitter validation webhook
+Should allow the VMI to have the new annotation only if:
+1. The feature gate is enabled on the Kubevirt CR
+2. VMI is configured with `isolateEmulatorThread: true`
```

Comment:
This can be checked down the stack and fail in the controller or while instantiating the VM. Consider removing this webhook

Reply:
I agree - it can, but it would require us to patch it. 
The reason the validating webhook was inserted was to only need to check the FG existence there, and make checking it later in the controller unnecessary, as advised by Lubo. I think we should settle it with him on the thread above first.

Reply:
Also note that rejecting the VMI here if the FG is disabled allows us to not check it on the virt-launcher. Not doing so requires us to pass the FG boolean via the GRPC. Not wanting to alter the GRPC with something the is alpha and might change in the future was also a reason to keep this webhook

Reply:
Is the only alternative checking it on virt-launcher? Can't you check it on virt-api controller level instead of the webhook?

Reply:
you have to check the annotation twice:
1. On the virt-controller, when setting extra CPUs for the emulator thread.
2. On the virt-launcher, when setting the CPUs for the emulatorThreadPin (will later set on the Housekeeping cgroup).

If we remove the validatin webhook - then we need to also check the FG on these occasions. As I mentioned - the accepted way to do this is the GRPC (see [example](https://github.com/kubevirt/kubevirt/blob/7e3aa9fff0226fea881fa4efc7906bad3cfcf4be/pkg/handler-launcher-com/cmd/v1/cmd.proto#L100)).

Reply:
IIUC saying "alpha annotation" implies that there is a graduation policy.
@xpivarc we need some clarifications here. Does an annotation require a FG?

Reply:
I would stay consistent because we never described the lifecycle of annotations :) We have a "stable" annotation and yet we do not have a graduation policy. For now, unfortunately, the only option that gates a behavior and is described are feature gates.